src/theory/strings/rewrites

```diff
@@ -193,15 +232,31 @@
 (define-rule* re-inter-flatten ((xs RegLan :list) (b RegLan) (ys RegLan :list) (zs RegLan :list)) (re.inter xs (re.inter b ys) zs) (re.inter xs b ys zs))
 (define-rule* re-inter-dup ((xs RegLan :list) (b RegLan) (ys RegLan :list) (zs RegLan :list)) (re.inter xs b ys b zs) (re.inter xs b ys zs))
 
-(define-rule* str-len-concat-rec ((s1 String) (s2 String) (s3 String :list))
-  (str.len (str.++ s1 s2 s3))
-  (str.len (str.++ s2 s3))
-  (+ (str.len s1) _))
-
-(define-cond-rule str-in-re-range-elim ((s String) (c1 String) (c2 String))
-  (and (= (str.len c1) 1) (= (str.len c2) 1))
-  (str.in_re s (re.range c1 c2))
-  (and (<= (str.to_code c1) (str.to_code s)) (<= (str.to_code s) (str.to_code c2))))
+(define-cond-rule re-inter-cstring ((xs RegLan :list) (ys RegLan :list) (s String))
+  (str.in_re s (re.inter xs ys))
+  (re.inter xs (str.to_re s) ys) 
+  (str.to_re s))
+
+(define-cond-rule re-inter-cstring-neg ((xs RegLan :list) (ys RegLan :list) (s String))
+  (not (str.in_re s (re.inter xs ys)))
+  (re.inter xs (str.to_re s) ys) 
+  re.none)
+
+(define-rule str-nth-elim-code ((s String) (n Int))
+  (seq.nth s n)
+  (str.to_code (str.substr s n 1)))
+
+(define-cond-rule str-len-substr-geq ((s String) (n Int) (m Int))
+  (>= m 0)
+  (>= (str.len (str.substr s m (str.len s))) n)
+  (>= (str.len s) (+ n m))
+)
```

Comment:
I could not quite see how this rule is correct. So, I tested it with IsaRARE and got the following counterexample:

```

m=1
n=0
s=""
```

Then,
`(>= m 0)` is true
`(str.len s)` is 0
`(str.substr s m (str.len s))` is "" since m > |s|
`(>= (str.len (str.substr s m (str.len s))) n)` is then true since 0>=n

However,
`(>= (str.len s) (+ n m))` is `(>= 0 (+ 0 1))` = `(>= 0 1)` and is false

Maybe there is a mistake within my reasoning but I think this is similiar to an error we had before where the special case of str.substr is not considered.

Reply:
Great find, I agree this rule is bugged.

For reference, I had been trying to get this rule to help express a rewrite that is now covered by an ad-hoc theory rewrite.  This is leftover from that attempt, but indeed is unsound and should be deleted.